src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+---
+title: 'Understanding the ECMAScript spec, part 1'
+author: '[Marja Hölttä](https://twitter.com/marjakh), speculative specification spectator'
+avatars:
+  - marja-holtta
+date: 2020-01-22 13:33:37
+tags:
+  - ECMAScript
+description: 'Tutorial on reading the ECMAScript specification'
+tweet: ''
+---
+
+... where we take a simple function in the spec and try to understand the notation.
+
+## Preface
+
+Even if you know JavaScript, reading its language specification, the ECMAScript spec, can be pretty daunting. At least that's how I felt when I started reading it for the first time. Reading the spec from the start doesn't help either, as the spec is not designed to be read in that order, and notations are not explained before they occur.
+
+Let's start with a concrete example and walk through the spec to understand it. The following code demonstrates usage of Object.prototype.hasOwnProperty:
+
+```javascript
+let o = {'foo': 1};
+o.hasOwnProperty('foo'); // true
+o.hasOwnProperty('bar'); // false
+```
+
+In the example, _o_ doesn't have a property called _hasOwnProperty_, so we walk up the prototype chain and look for it. We find it in _o_'s prototype, which is _Object.prototype_.
+
+When trying to find out how _Object.prototype.hasOwnProperty_ is defined, I bumped into pseudocode-like descriptions like this:
+
+> Object.prototype.hasOwnProperty(V)
+>
+> When the hasOwnProperty method is called with argument V, the following steps are taken:
+>
+> 1. Let P be **?** ToPropertyKey(V).
+> 2. Let O be **?** ToObject(this value).
+> 3. Return **?** HasOwnProperty(O, P).
+
+[Spec: Object.prototype.hasOwnProperty](https://tc39.es/ecma262#sec-object.prototype.hasownproperty)
+
+and
+
+> HasOwnProperty(O, P)
+>
+>The **abstract operation** HasOwnProperty is used to determine whether an object has an own property with the specified property key. A Boolean value is returned. The operation is called with arguments O and P where O is the object and P is the property key. This abstract operation performs the following steps:
+>
+> 1. **Assert**: Type(O) is Object.
+> 2. **Assert**: IsPropertyKey(P) is true.
+> 3. Let desc be **?** O.**\[\[GetOwnProperty\]\]**(P).
+> 4. If desc is undefined, return false.
+> 5. Return true.
+
+[Spec: HasOwnProperty](https://tc39.es/ecma262#sec-hasownproperty)
+
+But what's an "abstract operation"? What are the things inside \[\[ \]\]? Why is there a "?" in front of a function? What do the asserts mean?
+
+Let's find out!
```

Comment:
You could consider inserting a table of content here, what aspects of the spec you are going to describe here.

Reply:
We don't support ToCs yet (see #131) but we could of course manually insert a list.

Reply:
Undecided about this one...

Reply:
I'm deciding against a table of contents, I think it would break the flow; as all those terms would be just nonsense before they're explained.